token/services/tokens/tokens.go

```diff
@@ -0,0 +1,193 @@
+	}
+	if logger.IsEnabledFor(zapcore.DebugLevel) {
+		logger.Debugf("transaction [%s on (%s)] is known, extract tokens", txID, tms.ID())
+	}
+	request := tx.Request()
+	if request.Metadata == nil {
+		logger.Debugf("transaction [%s], no metadata found, skip it", txID)
+		return nil
+	}
+	metadata, err := request.GetMetadata()
+	if err != nil {
+		logger.Debugf("transaction [%s], failed to get metadata [%s]", txID, err)
+		return err
+	}
+
+	precision := tms.PublicParametersManager().PublicParameters().Precision()
+	if tms.PublicParametersManager().PublicParameters().GraphHiding() {
+		ids := metadata.SpentTokenID()
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("transaction [%s] with graph hiding, delete inputs [%v]", txID, ids)
+		}
+		for _, id := range ids {
+			if err := t.TokenStore.DeleteToken(id.TxId, id.Index, txID); err != nil {
+				return err
+			}
+		}
+	}
+
+	auditorFlag := t.Ownership.AmIAnAuditor(tms)
+	if auditorFlag {
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("transaction [%s], I must be the auditor", txID)
+		}
+	}
+
+	is, os, err := request.InputsAndOutputs()
+	if err != nil {
+		return errors.WithMessagef(err, "failed to get request's outputs")
+	}
+
+	// parse the inputs
+	for _, input := range is.Inputs() {
+		if input.Id == nil {
+			logger.Infof("transaction [%s] found an input that is not mine, skip it", tx.ID())
+			continue
+		}
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("transaction [%s] delete input [%s]", tx.ID(), input.Id)
+		}
+		if err := t.TokenStore.DeleteToken(input.Id.TxId, input.Id.Index, tx.ID()); err != nil {
+			return err
+		}
+		continue
+	}
+
+	// parse the outputs
+	for _, output := range os.Outputs() {
+		// get token in the clear
+		tok, issuer, tokenOnLedgerMetadata, err := metadata.GetToken(output.LedgerOutput)
+		if err != nil {
+			logger.Errorf("transaction [%s], found a token but failed getting the clear version, skipping it [%s]", txID, err)
+			continue
+		}
+		if tok == nil {
+			logger.Warnf("failed getting token in the clear for [%s, %s]", output.ID(txID), string(output.LedgerOutput))
+			continue
+		}
+
+		if len(output.LedgerOutput) == 0 {
+			if logger.IsEnabledFor(zapcore.DebugLevel) {
+				logger.Debugf("transaction [%s] without graph hiding, delete input [%d]", txID, output.Index)
+			}
+			if err := t.TokenStore.DeleteToken(tx.ID(), output.Index, tx.ID()); err != nil {
+				return err
+			}
+			continue
+		}
+
+		issuerFlag := !issuer.IsNone() && t.Issued.Issued(tms, issuer, tok)
+		ids, mine := t.Ownership.IsMine(tms, tok)
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			if mine {
+				logger.Debugf("transaction [%s], found a token and it is mine", txID)
+			} else {
+				logger.Debugf("transaction [%s], found a token and it is NOT mine", txID)
+			}
+		}
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
```

Comment:
I don't think we need this check if logger.IsEnabledFor if we don't do any calculations. But in any case we could at least put them all together.

Reply:
👍